content/blog/2019-10-21-threads-not-as-libraries/index.md

```diff
@@ -0,0 +1,272 @@
+
+To illustrate how register promotion can create a data race, consider the following example, where a shared variable `x` is modified:
+```c
+for (...) {
+  ...
+  if (mt) pthread_mutex_lock(...);
+  x = ... x...
+  if (mt) pthread_mutex_unlock(...);
+}
+```
+
+Suppose `mt` is only true when multiple threads have been created.
+Then this program lacks data races, since reads and writes to `x` are logically synchronized.
+
+Further suppose that register promotion is determined to be a beneficial optimization, perhaps because the conditional is observed to rarely be taken.
+Since the Pthreads library calls are simply regarded as opaque function calls, the optimization will produce the following:
+
+```c
+r = x; 
+for (...) {
+  ...
+  if (mt) {
+    x = r; pthread_mutex_lock(...); r = x;
+  }
+  r = ... r...
+  if (mt) {
+    x = r; pthread_mutex_unlock(...); r = x;
+  }
+}
+x = r; 
+```
+
+In this case, register promotion introduces unsynchronized reads and writes to `x`, creating a data race.
+
+
+## Allowing Data Races
+In the face of these issues, it’s clear that, indeed, threads cannot be implemented as a library. Instead, the thread specification must be formally incorporated into the language specification, to ensure that compilers do not break the former.
+
+The task of formally specifying Pthreads presents an opportunity for rethinking its synchronization-oriented paradigm.
+The author identifies that synchronization is not always desirable, and that it requires a precise interleaving of expensive atomic operations:
+>The cost of atomic operations and memory barriers varies widely, but is often comparable to that of a hundred or more register-to-register instructions, even in the absence of a cache miss. For example, on some Pentium 4 processors, hardware instructions to atomically update a memory location require well over 100 processor cycles, and these can also double as one of the cheaper mechanisms for ensuring that a store operation becomes visible to other threads before a subsequent load. 
+
+Instead of synchronization, the author argues in favor of a paradigm that allows data races and relies on atomic operations by showing performance gains from such a paradigm shift. 
+>As a result of the high cost of these hardware instructions, and the even higher cost of the pthread primitives built on them, there are a small number of cases in which synchronization performance is critical, and more careful and direct use of the hardware primitives, together with less constrained use of shared variables, is essential. In some cases it may also be necessary to avoid deadlock issues inherent in lock-based programming[7], or desirable because a different parallel programming model is preferable for an application (cf. [32]).
+
+To demonstrate this claim, the author compares the performance of an implementation of the Sieve of Eratosthenes algorithm and a tracing garbage collector implemented under the synchronization and data raceful paradigm.
```

Comment:
I don't know if I'd call it a "data raceful paradigm" that he's demonstrating here—he does use a racy version as a kind of "limit study," but the idea is that the programs are still correct because of careful use of atomic operations (for which the standard currently has no semantics).

Reply:
I'm not sure how else to describe it, and the paper seems to be hinting at that direction in different ways:
- Section 5.2 is called "Consequences of allowing data races". 

- This sentence from Section 5.2, which seems to hint at a future where we allow data races, or "concurrent updates":
> Once we allow concurrent updates, it falls on the language specification to give their semantics, and on the compiler itself to implement them.

- This excerpt seems to express a yearning for data races:
> In some cases it may also be necessary to avoid deadlock issues inherent in lock-based programming[7], or desirable because a different parallel programming model is preferable for an application (cf. [32])... The potential practical benefit of parallel algorithms involving races has long been recognized...


As I understand it, the author includes simultaneous atomic reads and writes under the term "data race", which is confusing/possibly incorrect. (Perhaps he was going for "race condition" instead?)

Reply:
Super good point. It is indeed a terminology shift. In 2019, we would not call a program that uses atomics to enforce parallel correctness "racy." That would contravene the "DRF=>SC" slogan—racy programs are never considered correct! That is, atomics are a way to avoid data races, just like locks. But obviously that terminology was not in place by 2004.